bundle/jsky.app.ot/src/main/scala/edu/gemini/catalog/ui/image/BackgroundImageLoader.scala

```diff
@@ -0,0 +1,220 @@
+      val name = s"Background Image Downloads - ${threadNumber.getAndIncrement()} - priority: $priority"
+      defaultThreadFactory.newThread(r) <| {_.setDaemon(true)} <| {_.setName(name)} <| {_.setPriority(priority)}
+    }
+  }
+
+  /**
+    * Execution context for lower priority downloads
+    */
+  private val lowPriorityEC = Executors.newFixedThreadPool(ImageCatalog.all.length, imageDownloadsThreadFactory(Thread.MIN_PRIORITY))
+
+  /**
+    * Regular execution context for higher priority tasks, i.e. UI requests
+    */
+  private val highPriorityEC = Executors.newFixedThreadPool(ImageCatalog.all.length, imageDownloadsThreadFactory(Thread.NORM_PRIORITY))
+
+  /** Called when a program is created to download its images */
+  def watch(prog: ISPProgram): Unit = {
+    // Listen for future changes
+    prog.addCompositeChangeListener(ChangeListener)
+    val targets = prog.getAllObservations.asScala.toList.flatMap(_.getObsComponents.asScala.flatMap(n => requestedImage(TpeContext(n))))
+    // remove duplicates and request images
+    val tasks = targets.distinct.map(requestImageDownload(lowPriorityEC))
+    // Run as low priority
+    runAsync(tasks)(_ => ())(lowPriorityEC)
+  }
+
+  /** Called when a program is removed to clear the cache */
+  def unwatch(prog: ISPProgram): Unit = {
+    prog.removeCompositeChangeListener(ChangeListener)
+  }
+
+  /**
+    * Display an image if available on disk or request the download if necessary
+    */
+  def loadImageOnTheTpe(tpe: TpeContext): Unit = {
+    val task = requestedImage(tpe).fold(taskUnit)(requestImageDownload(highPriorityEC))
+
+    // This method called on an explicit user interaction so we'd rather
+    // Request the execution in a higher priority thread
+    // Execute and set the image on the tpe
+    runAsync(task)(_ => ())(highPriorityEC)
+  }
+
+  // Watches for changes to existing observations, runs BAGS on them when updated.
+  object ChangeListener extends PropertyChangeListener {
+    override def propertyChange(evt: PropertyChangeEvent): Unit =
+      evt.getSource match {
+        case node: ISPNode => Option(node.getContextObservation).foreach { o =>
+          val task = requestedImage(TpeContext(o)).fold(taskUnit)(requestImageDownload(highPriorityEC))
+
+          // Run it in the background as it is lower priority than GUI
+          runAsync(task)(_ => ())(lowPriorityEC)
+        }
+      }
+  }
+
+  /**
+    * Creates a task to load an image and set it on the tpe
+    */
+  private[image] def requestImageDownload(pool: ExecutorService)(t: TargetImageRequest): Task[Unit] =
+    for {
+      catalog  <- ObservationCatalogOverrides.catalogFor(t.key, t.obsWavelength)
+      image    <- loadImage(ImageSearchQuery(catalog, t.coordinates, catalog.imageSize), ImageCatalogPanel.resetListener)(pool)
+    } yield image match {
+      case Some(e) => updateTpeImage(e) // Try to set it on the UI
+      case _       => // Ignore, some other thread is downloading the image
+    }
```

Comment:
`updateTpeImage` is a side-effect so you need to capture it with `delay`. Something like

```scala
for {
  catalog  <- ObservationCatalogOverrides.catalogFor(t.key, t.obsWavelength)
  image    <- loadImage(ImageSearchQuery(catalog, t.coordinates, catalog.imageSize), ImageCatalogPanel.resetListener)(pool)
  _        <- image.fold(taskUnit)(e => Task.delay(updateTpeImage(e)))
} yield ()
```